File: app/src/main/java/ua/hospes/rtm/utils/Optional.java
package ua.hospes.rtm.utils;

import androidx.annotation.Nullable;

/**
 * @author Andrew Khloponin
 */
public class Optional<T> {
    @Nullable
    private T t = null;


    private Optional() {}


    public static <T> Optional<T> empty() {
        return new Optional<>();
    }

    public static <T> Optional<T> of(T value) {
        Optional<T> o = new Optional<>();
        o.t = value;
        return o;
    }


    @Nullable
    public T get() {
        return t;
    }

    public boolean isPresent() {
        return t != null;
    }
}

File: app/src/main/java/ua/hospes/rtm/utils/Optional.kt
package ua.hospes.rtm.utils

/**
 * @author Andrew Khloponin
 */
class Optional<T> private constructor() {
    private var t: T? = null

    val isPresent: Boolean
        get() = t != null


    fun get(): T? {
        return t
    }

    companion object {


        fun <T> empty(): Optional<T> {
            return Optional()
        }

        fun <T> of(value: T): Optional<T> {
            val o = Optional<T>()
            o.t = value
            return o
        }
    }
}